shell/platform/embedder/embedder_external_view_embedder.cc

```diff
@@ -0,0 +1,277 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/embedder/embedder_external_view_embedder.h"
+
+#include <algorithm>
+
+#include "flutter/shell/platform/embedder/embedder_render_target.h"
+
+namespace flutter {
+
+EmbedderExternalViewEmbedder::EmbedderExternalViewEmbedder(
+    CreateRenderTargetCallback create_render_target_callback,
+    PresentCallback present_callback)
+    : create_render_target_callback_(create_render_target_callback),
+      present_callback_(present_callback) {
+  FML_DCHECK(create_render_target_callback_);
+  FML_DCHECK(present_callback_);
+}
+
+EmbedderExternalViewEmbedder::~EmbedderExternalViewEmbedder() = default;
+
+void EmbedderExternalViewEmbedder::Reset() {
+  pending_recorders_.clear();
+  pending_params_.clear();
+  composition_order_.clear();
+}
+
+// |ExternalViewEmbedder|
+bool EmbedderExternalViewEmbedder::HasPendingViewOperations() {
+  // This is used to decide whether to switch thread configurations. Since the
+  // embedder API does not support dynamic thread configurations, we never want
+  // to indicate (even accidentally) that this is true. Currently, it is
+  // embedder responsibility to flush pending view operations to their hierarchy
+  // in the compositor present callback.
+  return false;
+}
+
+// |ExternalViewEmbedder|
+void EmbedderExternalViewEmbedder::CancelFrame() {
+  Reset();
+}
+
+static FlutterBackingStoreConfig MakeBackingStoreConfig(const SkISize& size) {
+  FlutterBackingStoreConfig config = {};
+
+  config.struct_size = sizeof(config);
+
+  config.size.width = size.width();
+  config.size.height = size.height();
+
+  return config;
+}
+
+// |ExternalViewEmbedder|
+void EmbedderExternalViewEmbedder::BeginFrame(SkISize frame_size,
+                                              GrContext* context) {
+  Reset();
+  pending_frame_size_ = frame_size;
+
+  // Decide if we want to discard the previous root render target.
+  if (root_render_target_) {
+    auto surface = root_render_target_->GetRenderSurface();
+    // This is unlikely to happen but the embedder could have given us a render
```

Comment:
Please be more specific about 'us', 'we', etc. (go/avoidwe)

Reply:
Done.